pkg/broker/api.go

```diff
@@ -177,82 +177,83 @@ func (b *AwsBroker) Provision(request *osb.ProvisionRequest, c *broker.RequestCo
-		glog.Errorln(errmsg)
-		r.LastOperationResponse.State = "failed"
-		r.LastOperationResponse.Description = &errmsg
-		return &r, nil
-	}
-	glog.V(10).Infoln(si.Params)
-	cfnsvc := b.Clients.NewCfn(b.GetSession(b.keyid, b.secretkey, b.region, b.accountId, b.profile, si.Params))
-	response, err := cfnsvc.Client.DescribeStacks(&cloudformation.DescribeStacksInput{StackName: aws.String(si.StackID)})
+		desc := fmt.Sprintf("Failed to get the service instance %s: %v", request.InstanceID, err)
+		return nil, newHTTPStatusCodeError(http.StatusInternalServerError, "", desc)
+	} else if instance == nil {
+		// Returning 410 Gone here is only appropriate for asynchronous delete
+		// operations, so hopefully that's what this operation is :)
+		// (https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md#response-1)
+		desc := fmt.Sprintf("The service instance %s was not found.", request.InstanceID)
+		return nil, newHTTPStatusCodeError(http.StatusGone, "", desc)
+	}
+
+	// Get the CFN stack status
+	cfnSvc := b.Clients.NewCfn(b.GetSession(b.keyid, b.secretkey, b.region, b.accountId, b.profile, instance.Params))
+	resp, err := cfnSvc.Client.DescribeStacks(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(instance.StackID),
+	})
 	if err != nil {
-		panic(err)
-	}
-	failedstates := []string{"CREATE_FAILED", "ROLLBACK_IN_PROGRESS", "ROLLBACK_FAILED", "ROLLBACK_COMPLETE", "DELETE_FAILED", "UPDATE_ROLLBACK_IN_PROGRESS", "UPDATE_ROLLBACK_FAILED", "UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS", "UPDATE_ROLLBACK_COMPLETE"}
-	progressingstates := []string{"CREATE_IN_PROGRESS", "DELETE_IN_PROGRESS", "UPDATE_IN_PROGRESS", "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS"}
-	successfulstates := []string{"CREATE_COMPLETE", "DELETE_COMPLETE", "UPDATE_COMPLETE"}
-	status := *response.Stacks[0].StackStatus
-	if stringInSlice(status, failedstates) {
-		glog.Errorf("CloudFormation stack failed %#+v", si.StackID)
-		glog.Errorf(status)
-		r.LastOperationResponse.State = "failed"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
-	} else if stringInSlice(status, progressingstates) {
-		glog.Infoln("CloudFormation stack still busy...")
-		glog.Infoln(status)
-		r.LastOperationResponse.State = "in progress"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
-	} else if stringInSlice(status, successfulstates) {
-		glog.Infoln("CloudFormation stack operation completed...")
-		glog.Infoln(status)
-		r.LastOperationResponse.State = "succeeded"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
+		desc := fmt.Sprintf("Failed to describe the CloudFormation stack %s: %v", instance.StackID, err)
+		return nil, newHTTPStatusCodeError(http.StatusInternalServerError, "", desc)
+	}
+
+	response := broker.LastOperationResponse{}
+	stackStatus := aws.StringValue(resp.Stacks[0].StackStatus)
+	if strings.HasSuffix(stackStatus, "_IN_PROGRESS") {
```

Comment:
Rather than hardcoding all of [the possible stack status codes](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html#w2ab1c15c15c17c11) here, what do you think of this approach: any `_IN_PROGRESS` status is considered in progress; `CREATE_COMPLETE`, `UPDATE_COMPLETE`, and `DELETE_COMPLETE` are considered successful; and anything else is considered a failure.

Reply:
downside I see of this is that we could fail faster for the various rollback in progress states (as they will always result in a failed state)